src/stdio.rs

```diff
@@ -0,0 +1,131 @@
+//! Enables direct write access to I/O streams in Python's `sys` module.
+
+//! In some cases printing to Rust's `std::io::stdout` or `std::io::stderr` will not appear
+//! in the Python interpreter, e.g. in Jupyter notebooks. This module provides a way to write
+//! directly to Python's I/O streams from Rust in such cases.
+
+//! ```rust
+//! let mut stdout = pyo3::stdio::stdout();
+//!   
+//! // This may not appear in Jupyter notebooks...
+//! println!("Hello, world!");
+//!
+//! // ...but this will.
+//! writeln!(stdout, "Hello, world!").unwrap();
+//! ```
+
+use crate::ffi::{PySys_WriteStderr, PySys_WriteStdout};
+use crate::prelude::*;
+use std::io::{LineWriter, Write};
+use std::marker::PhantomData;
+use std::os::raw::{c_char, c_int};
+
+trait PyStdioRawConfig {
+    const STREAM: &'static str;
+    const PRINTFCN: unsafe extern "C" fn(*const i8, ...);
+}
+
+struct PyStdoutRaw {}
+impl PyStdioRawConfig for PyStdoutRaw {
+    const STREAM: &'static str = "stdout";
+    const PRINTFCN: unsafe extern "C" fn(*const i8, ...) = PySys_WriteStdout;
+}
+
+struct PyStderrRaw {}
+impl PyStdioRawConfig for PyStderrRaw {
+    const STREAM: &'static str = "stderr";
+    const PRINTFCN: unsafe extern "C" fn(*const i8, ...) = PySys_WriteStderr;
+}
+
+struct PyStdioRaw<T: PyStdioRawConfig> {
+    pystream: Py<PyAny>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: PyStdioRawConfig> PyStdioRaw<T> {
+    fn new() -> Self {
+        let pystream: Py<PyAny> = Python::with_gil(|py| {
+            let module = PyModule::import_bound(py, "sys").unwrap();
+            module.getattr(T::STREAM).unwrap().into()
+        });
+
+        Self {
+            pystream,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T: PyStdioRawConfig> Write for PyStdioRaw<T> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        Python::with_gil(|_py| unsafe {
+            (T::PRINTFCN)(
+                b"%.*s\0".as_ptr().cast(),
+                buf.len() as c_int,
+                buf.as_ptr() as *const c_char,
+            );
+        });
+        Ok(buf.len())
+    }
+    fn flush(&mut self) -> std::io::Result<()> {
+        Python::with_gil(|py| -> std::io::Result<()> {
+            self.pystream
+                .call_method0(py, "flush")
```

Comment:
We should [`intern!`](https://docs.rs/pyo3/latest/pyo3/macro.intern.html) the name "flush" here.

Reply:
Done.